map.go

```diff
@@ -387,6 +387,88 @@ func (spec *MapSpec) createMap(inner *sys.FD, opts MapOptions) (_ *Map, err erro
 	if err != nil {
 		return nil, err
 	}
+	switch spec.Type {
```

Comment:
Even if this change is still in draft mode, I want to mention the point, that map flags are subject of change. So hardcoding flag checks might leed to confusion or limit usages, especially if the version/patch status of the kernel is unknown.

[This is an example](https://github.com/torvalds/linux/commit/591fe9888d7809d9ee5c828020b6c6ae27c37229) for flag requirements for newer kernel.

What about providing hints based on the flag, if the map creation fails?

Reply:
I agree. I think @lmb also mentioned it could be a pain to maintain although it does help for better UX. I was thinking of a simple hint like:
```
if errors.Is(err, unix.EINVAL) && spec.Flags&unix.BPF_F_NO_PREALLOC > 0{
      return nil, fmt.Errorf("map create: %w (noPrealloc flag may be incompatible with map type %s)", err, spec.Type)
}
```

Reply:
Agree with Florian, this must go into the `if err != nil {` case just after `MapCreate` (for the same reason we put the other feature detection checks there). Pulling it out into a separate function would also be nice.